### ISSUE TITLE
Problem (?) with EXTRA in import_name and import_from

### ISSUE THREAD

**Issue by lysnikolaou** [gvanrossum/pegen#72]
Going through all the things I did for #67 and #71, I realised that there might be a slight problem we didn't catch, but I am not sure.

Let me explain with the example of the `import_name` rule:
```
import_name[stmt_ty]: a='import' b=dotted_as_names { _Py_Import(b, EXTRA(a, b)) }
```

`b` is an `asdl_seq *` and we use it as the tail parameter to `EXTRA`, which calls `ENDLINE` and `ENDCOL` for `b`, which then access the `end_lineno` and `end_col_offset` attributes after casting the `asdl_seq *` to an `expr_ty`. I don't really see how that doesn't generate a SEGFAULT, but (SEGFAULTS aside) I think it will probably be a bug in the future, when these AST nodes are used for error reporting.

Am I maybe missing something?

**Comment by gvanrossum** [comment]
Ouch, definitely a real problem. Good catch! Not sure what to do, maybe a
helper function?
-- 
--Guido (mobile)

**Comment by lysnikolaou** [comment]
I think that a helper function is the way to go as well. Something like `asdl_seq_get_tail_elem()`?

**Comment by gvanrossum** [comment]
Yeah, though what to do if the length is zero? Maybe pass the previous item
in as well?

We should also try to make it more likely that this is caught by the C type
checker.
-- 
--Guido (mobile)

**Comment by lysnikolaou** [comment]
> We should also try to make it more likely that this is caught by the C type checker.

The problem with that is that the we need the cast to `expr_ty`, which makes it possible to pass in whatever pointer type we want and the C type checker will never say anything about it.

**Comment by gvanrossum** [comment]
Huh, does only expr_ty have line/column info?

**Comment by lysnikolaou** [comment]
No, but, to the best of my knowledge, the only general types that do are `expr_ty` and `stmt_ty` (there is also `arg_ty` and `excepthandler_ty`, but those aren't general enough for us to cast to). The problem is that an `asdl_seq *` can/will hold `void *`s, which means that we _need_ to cast to either of those, in order to be able to access line/column info (I think it doesn't really matter to which exactly). Due to this we can't really get much help from the type checker.

We could cast in the helper function though, where we know if the element comes from an `asdl_seq *` or not, but the type checker still doesn't come into play.

**Comment by gvanrossum** [comment]
Hm, yeah, C doesn't have generics... :-(

I guess we'll have to just live dangerously and document what we're doing.

Anyway, in order to test that we're producing the right line/col info, we should use `ast.dump(..., include_attributes=True)` in our tests. That should reveal bogus line/col info.

**Comment by lysnikolaou** [comment]
Ok then, let's do that. I would like to start documenting the work I'm doing. Do you have a recommendation as to what format to use? Just text/markdown files or should we setup something like sphinx?

**Comment by gvanrossum** [comment]
I like markdown for this purpose. And agreed that it’s important to write
things down!

> --
--Guido (mobile)

**Comment by lysnikolaou** [comment]
Turns out casting to either `stmt_ty` or `expr_ty` doesn't work, since the union in those could take up a different number of bytes, which leads to the line/col fields to have different offsets to the base of the struct. We need to somehow figure out what we are passing to EXTRA and cast explicitly to that.

There doesn't seem to be an obvious way of doing this, or at least not to me. This also seems like an important design decision, because if we get it wrong, then the whole grammar will need to be refactored again, which won't be trivial, when we have lots of actions implemented.

**Comment by gvanrossum** [comment]
Hm, so what we may be able to do is keep the one `EXTRA(head, tail)` macro, and ensure that the arguments have the right type. E.g. we could have a helper function to extract the last `expr_ty` from an `asdl_seq *` and another helper to extract the last `stmt_ty`. (These could be the same function, returning a `void *`, and two different macros to cast to `expr_ty` and `stmt_ty`.)

Currently it looks like `EXTRA()` won't work for tokens at all, but we can make it work by renaming the relevant fields in the `Token` struct to match what `EXTRA()` uses.

It also looks like the `(expr_ty)` casts need to be removed from `LINE()`, `COL()`, `ENDLINE()` and `ENDCOL()`. This probably means that in those places where we currently pass a `void *` into `EXTRA()` we need to cast it explicitly to the right type. (I'd add a macro for that, since C's cast notation is pretty ugly.)

It does require that all the `asdl_seq` objects we create are uniform (meaning that the items are all `expr_ty`, or all `stmt_ty`, etc.), but that does not look like a great burden. I expect we won't need to have arrays of tokens? Or maybe only one case.

**Comment by lysnikolaou** [comment]
I think I might have a better approach, which I am currently working on. I'll have an update as to if it works in the next half hour.

**Comment by lysnikolaou** [comment]
So I have an update on this: It works! I don't know, it is okay with you, but this was the best approach I could come up with and I think it is easier to main and extend:

We get rid of all the macros instead of EXTRA, which gets a whole new definition:
```
#define EXTRA(head, head_type_func, tail, tail_type_func) head_type_func(head, 1, 1), \
                                                          head_type_func(head, 1, 0), \
                                                          tail_type_func(tail, 0, 1), \
                                                          tail_type_func(tail, 0, 0), \
                                                          p->arena
```

Furthermore, we define three functions, one for each type:
```
int expr_type(void *, int, int);
int stmt_type(void *, int, int);
int token_type(void *, int, int);
```

All of them do the same things, up to casting to the correct type. For example:
```
int
expr_type(void *a, int head, int line)
{
    if (head && line)
        return ((expr_ty) a)->lineno;
    else if (head && !line)
        return ((expr_ty) a)->col_offset;
    else if (!head && line)
        return ((expr_ty) a)->end_lineno;
    else
        return ((expr_ty) a)->end_col_offset;
}
```

This way, we can just use something like `EXTRA(a, token_type, b, stmt_type)`, which seems like a good trade-off. We do not need to change the definition of `Token` and all we have to do to extend it for new types is define a new function that casts to it.

What do you think?

**Comment by gvanrossum** [comment]
This looks like the way to go. I think you should make a PR for this!

I'm worried a tiny bit about having EXTRA() take 4 arguments instead of 2, but we can always have a few more macros to handle common cases. I agree we can't have a separate macro for each combination of head and tail type.

In any case I think we should plan to occasionally change how EXTRA() is implemented but we can keep its interface the same going forward.

PS. I may not be available the rest of the day.

**Comment by lysnikolaou** [comment]
Fixed through #73.

**Comment by gvanrossum** [comment]
PS. If you include "Fixes #72" somewhere in the test of the commit message,
merging the PR will auto-close the issue.
-- 
--Guido (mobile)
